crates/core/src/ring/location.rs

```diff
@@ -0,0 +1,283 @@
+        Self::from_contract_key(key.as_bytes())
+    }
+}
+
+impl Display for Location {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl PartialEq for Location {
+    fn eq(&self, other: &Self) -> bool {
+        (self.0 - other.0).abs() < f64::EPSILON
+    }
+}
+
+/// Since we don't allow NaN values in the construction of Location
+/// we can safely assume that an equivalence relation holds.  
+impl Eq for Location {}
+
+impl Ord for Location {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0
+            .partial_cmp(&other.0)
+            .expect("always should return a cmp value")
+    }
+}
+
+impl PartialOrd for Location {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl std::hash::Hash for Location {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let bits = self.0.to_bits();
+        state.write_u64(bits);
+    }
+}
+
+impl TryFrom<f64> for Location {
+    type Error = anyhow::Error;
+
+    fn try_from(value: f64) -> Result<Self, Self::Error> {
+        if !(0.0..=1.0).contains(&value) {
+            bail!("expected a value between 0.0 and 1.0, received {}", value)
+        } else {
+            Ok(Location(value))
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Distance(f64);
+
+impl Distance {
+    pub fn new(value: f64) -> Self {
+        debug_assert!(!value.is_nan(), "Distance cannot be NaN");
+        debug_assert!(
+            (0.0..=1.0).contains(&value),
+            "Distance must be in the range [0, 1.0]"
+        );
+        if value <= 0.5 {
+            Distance(value)
+        } else {
+            Distance(1.0 - value)
+        }
+    }
+
+    pub fn as_f64(&self) -> f64 {
+        self.0
+    }
+}
+
+impl Add for Distance {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let d = self.0 + rhs.0;
+        if d > 0.5 {
+            Distance::new(1.0 - d)
+        } else {
+            Distance::new(d)
+        }
+    }
+}
+
+impl PartialEq for Distance {
+    fn eq(&self, other: &Self) -> bool {
+        (self.0 - other.0).abs() < f64::EPSILON
+    }
+}
+
+impl PartialOrd for Distance {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for Distance {}
+
+impl Ord for Distance {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0
+            .partial_cmp(&other.0)
+            .expect("always should return a cmp value")
+    }
+}
+
+impl Display for Distance {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_ipv4_address_location() {
+        use rand::Rng;
+        use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+        let mut rng = rand::thread_rng();
```

Comment:
for testing i think is always good to use a PRNG so you can have reproducible failures (can do multiple rounds if you want to cover more samples)